lib/protocol/protocol.go

```diff
@@ -363,25 +370,44 @@ func (c *rawConnection) ping() bool {
 	return c.send(&Ping{}, nil)
 }
 
-func (c *rawConnection) readerLoop() (err error) {
-	defer c.wg.Done()
+type errorMessage struct {
+	msg message
+	err error
+}
+
+func (c *rawConnection) readerLoop() error {
 	fourByteBuf := make([]byte, 4)
+	inbox := make(chan errorMessage)
+
+	// Reading from the wire may block until the underlying connection is closed.
+	go func() {
+		for {
+			msg, err := c.readMessage(fourByteBuf)
+			select {
+			case inbox <- errorMessage{msg: msg, err: err}:
+			case <-c.closed:
+				return
+			}
+		}
+	}()
+
 	state := stateInitial
+	var errMsg errorMessage
 	for {
-		if c.Closed() {
+		select {
+		case errMsg = <-inbox:
+		case <-c.closed:
```

Comment:
Shouldn't we just wait for readMessage to return ErrClosed or whatever, to make sure it exits?
Also, who closes inbox?

Reply:
That's exactly what we can't do: If e.g. the model closes the connection, `readMessage` won't return until `internalClose`/`wg.Wait` returns. Ideally `io.Reader` would have an equivalent with channels, but unfortunately it doesn't.
And nobody closes `inbox` currently, I can make the `readMessage` goroutine close it, but it's irrelevant as the receiver already terminated at that point.